### PR TITLE
Allow customers to add a custom state to federatedSignIn method

### LINKED ISSUE #3125: OAuth federation: Redirect User to the last requested page after authentication
** Which Category is your question related to? **

Amplify Auth Module

** What AWS Services are you utilizing? **

Cognito

** Provide additional details e.g. code snippets **
I'm currently using OAuth support for Amplify that was recently released. The challenge I’m running into with this new flow is redirecting a user to the last page they requested. I learned that Cognito app-client doesn’t allow for wildcard entries. 

In this instance, I'm using `Auth.federatedSignIn({ provider: 'Google' })`.When a user tries to access `https://example.com?search=<random-guid>` in the unauthenticated state, the user gets presented with google prompt, after signing in, the user is redirected back to a `redirectSignIn` url, which is a static location `https://example.com`. I would instead want the user to be redirected back to the last page they requested, which is a `https://example.com?search=<random-guid>`. I also learned that there is a way to pass a `state` variable to Cognito, but struggling to find a way to do that in the using Amplify. Can you please provide a code snippet?

I'm using the authenticator component `<Authenticator
        theme={authenticatorTheme}
        federated={federated}
        hide={[Greetings]}
      />`
What I noticed is, this component with `Auth.configure({ oauth })` adds "Sign In with AWS" button (Hosted UI) to the SignIn form. How do I hide that without hiding the whole SignIn component?

Can you please help?

Thanks,
Rohit

### LINKED ISSUE #3177: React Native federatedSignin with Google is not working
I cannot get Google signin to work on react native with Auth.federatedSignin({provider:"Google"});

We have what I think, is a very common scenario: Our requirement is to allow google/facebook signin for web and react native. We are using Cognito Userpools. 
Starting with web, I added "https://mydomain.io/login/" for signin in our userpool client, got client creds from google, and fed them to the Google Identity Provider in the userpool. This setup works just fine - I'm able to signin and create new users, and even invoke api in api gateway just fine.

For react native, we setup url scheme "myapp" in Info.plist for iOS to enable deep linking. I also added "myapp://authorize/" to the signin redirects in userpool app client. So our Callback URL in UserPool looks like "myapp://authorize/, "https://mydomain.io/login/". I then call Auth.federatedSignin({provider:"Google"}) and setup Hub listeners for the auth channel. On making this call, the client opens the browser, shows the hosted UI, signs the user in and returns back to our app. However, things just break down from there. 

This is what I see in the logs:
```
DEBUG] 21:46.677 OAuth - Starting code flow with myapp://authorize/?code=guid&state=longstring
[DEBUG] 21:46.678 Hub - Dispatching to auth with {"event":"codeFlow","data":{},"message":"Retrieving tokens from https://mydomain.auth.us-west-2.amazoncognito.com/oauth2/token"}
[DEBUG] 21:46.679 Analytics - on hub capsule auth{"event":"codeFlow","data":{},"message":"Retrieving tokens from https://mydomain.auth.us-west-2.amazoncognito.com/oauth2/token"}
[DEBUG] 21:46.680 Hub - Dispatching to auth with {"event":"codeFlow","data":{},"message":"Retrieving tokens from https://mydomain.auth.us-west-2.amazoncognito.com/oauth2/token"}
{"channel":"auth","payload":{"event":"codeFlow","data":{},"message":"Retrieving tokens from https://mydomain.auth.us-west-2.amazoncognito.com/oauth2/token"},"source":"Auth","patternInfo":[]}
{"event":"codeFlow","data":{},"message":"Retrieving tokens from https://mydomain.auth.us-west-2.amazoncognito.com/oauth2/token"}
[DEBUG] 21:46.683 OAuth - Calling token endpoint: https://mydomain.auth.us-west-2.amazoncognito.com/oauth2/token with{"grant_type":"authorization_code","code":"guid","client_id":"7pa0s7cc3tpolm26g3eno9g04r","redirect_uri":"myapp://authorize/","code_verifier":"long string"}
[DEBUG] 21:46.869 AuthClass - Error in cognito hosted auth response{}
[DEBUG] 21:46.870 Hub - Dispatching to auth with {"event":"signIn_failure","data":{},"message":"The OAuth response flow failed"}
[DEBUG] 21:46.874 Analytics - on hub capsule auth{"event":"signIn_failure","data":{},"message":"The OAuth response flow failed"}
[DEBUG] 21:46.875 Credentials - getting credentials
[DEBUG] 21:46.876 Credentials - picking up credentials
[DEBUG] 21:46.878 Credentials - getting new cred promise
[DEBUG] 21:46.879 Credentials - checking if credentials exists and not expired
[DEBUG] 21:46.880 Credentials - need to get a new credential or refresh the existing one
[DEBUG] 21:46.880 AuthClass - Getting current user credentials
[DEBUG] 21:46.881 Hub - Dispatching to auth with {"event":"signIn_failure","data":{},"message":"The OAuth response flow failed"}
[DEBUG] 21:46.882 Hub - Dispatching to auth with {"event":"signIn_failure","data":{},"message":"The OAuth response flow failed"}
{"channel":"auth","payload":{"event":"signIn_failure","data":{},"message":"The OAuth response flow failed"},"source":"Auth","patternInfo":[]}
{"event":"signIn_failure","data":{},"message":"The OAuth response flow failed"}
[DEBUG] 21:46.883 Hub - Dispatching to auth with {"event":"cognitoHostedUI_failure","data":{},"message":"A failure occurred when returning to the Cognito Hosted UI"}
[DEBUG] 21:46.883 Analytics - on hub capsule auth{"event":"cognitoHostedUI_failure","data":{},"message":"A failure occurred when returning to the Cognito Hosted UI"}
[DEBUG] 21:46.884 Hub - Dispatching to auth with {"event":"cognitoHostedUI_failure","data":{},"message":"A failure occurred when returning to the Cognito Hosted UI"}
{"channel":"auth","payload":{"event":"cognitoHostedUI_failure","data":{},"message":"A failure occurred when returning to the Cognito Hosted UI"},"source":"Auth","patternInfo":[]}
{"event":"cognitoHostedUI_failure","data":{},"message":"A failure occurred when returning to the Cognito Hosted UI"}
```

This is what my amplify.configure looks:
```
{"Auth": {
		"userPoolId": "userpoolid",
		"identityPoolId": "identitypoolid",
		"region": "us-west-2",
		"userPoolWebClientId": "webclientid",
		"oauth": {
			"domain": "mydomain.auth.us-west-2.amazoncognito.com",
			"redirectSignIn": "myapp://authorize/",
			"redirectSignOut": "myapp://logout/",
			"scope": [
				"profile",
				"openid",
				"email"
			],
			"responseType": "code"
		}
	}
}
```

Is there any documentation I can look at to help me unblock myself?

### LINKED ISSUE #3188: Using Hosted UI with Ionic / Cordova application
It is currently possible but difficult to integrate Ionic and Cordova based applications with Amplify Auth using HostedUI to allow social login.  The scenario I needed to implement was social authentication with Facebook and username/password authentication on Android / IOS. Where as integrating into a browser based application is relatively straightforward the documentation around mobile integration is sparse to say the least. To get the this to work with the amplify ionic integration I needed to 

* Call `Auth.federatedSignIn({provider: CognitoHostedUIIdentityProvider.Facebook })` to launch the hosted UI in the mobile browser and set a custom URL scheme for the callback (i.e. myapp://auth/)

* Use the Deeplink cordova plugin (https://github.com/ionic-team/ionic-plugin-deeplinks) to intercept custom url schemes in application

This correctly launches the hosted UI, allows facebook authentication and then returns to the application with the URL containing 'code' and 'state parameters.  Then in a deeplink handler I then can pass this return URL to complete this part of the flow using `;(Auth as any)._handleAuthResponse(deeplink.$link.url)`.

At this point it will handle the auth response and exchange the code for a token and then try to redirect to the `redirectSignIn` url as before and this will throw an exception

`Error: Failed to execute 'replaceState' on 'History': A history state object with URL 'myapp://app/auth/' cannot be created in a document with origin` from Auth.ts:1553

this can also be worked around by just catching this exception and then retrieving the user credentials.

**What would be lovely**
It would be nice to be able to 

* avoid the need to use the any cast to call into _handleAuthResponse with a first class public function on the Auth class to perform this duty.

* allow specifying a seperate URL for internal redirection after auth flow is complete that would override  `redirectSignIn` parameter for this if specified

* improve the documentation to explain how to implement this in more detail

I am happy to help contributing patches for any of these if there is agreement to improve the situation here and the best approach to it.

### LINKED ISSUE #3194: How to pass query params with oauth configuration.
Hello, 
I am trying to do faderatedSignin by Google/Facebook using oauth and hosted UI.
Everything is working fine but after successful signin it's redirecting back to my redirect url with no query params.

For example, I have signin page with url "http://localhost:3000/?name=xyz". In the signin component I have added two buttons for Google and Facebook and calling faderatedSignin function with provider. After that it's redirecting to google/facebook page and I am able to do signin through it. After successful signin it's redirect back to "http://localhost:3000/". But I am not able to access query param (?name=xyz). Please check the below code and do the needful.

OAuth configuration:
```
const awsmobile = {
    "aws_project_region": "XX",
    "aws_cognito_region": "XX",
    "aws_user_pools_id": "XX",
    "aws_user_pools_web_client_id": "XX",
    "oauth": {
        "domain": "XX.amazoncognito.com",
        "scope": [
            "email",
            "openid",
            "profile",
            "aws.cognito.signin.user.admin"
        ],
        "redirectSignIn": "http://localhost:3000/",
        "redirectSignOut": "http://localhost:3000/",
        "responseType": "code"
    },
    "federationTarget": "COGNITO_USER_POOLS"
};
```
SignIn Component:
```
import React, { Component } from 'react';
import Amplify, { Auth, Hub } from 'aws-amplify';
import awsmobile from './aws-exports';
Amplify.configure(awsmobile);

class SignIn extends Component {
  
      constructor(props) {
          super(props);
  
          Hub.listen("auth", ({ payload: { event, data } }) => {
            console.log('data:---', data);
            switch (event) {
              case "signIn":
                  console.log('signin:---');
                  break;
              case "signOut":
                  console.log('signout:---');
                  break;
            }
          });
          this.state = {
             
          };
  
      }
  
  
      render() {
        
              return (
                  <div>
                      
                        <button onClick={() => Auth.federatedSignIn({provider: 'Google'})}>Open Google</button>
                        <button onClick={() => Auth.federatedSignIn({provider: 'Facebook'})}>Open Facebook</button>
                        <button onClick={() => Auth.federatedSignIn()}>Open Hosted UI</button>
                        <button onClick={() => Auth.signOut()}>Sign out</button>
                   
                  </div>
              );
          }
        
      }
  

export default SignIn;
```

### LINKED ISSUE #3569: How to set STATE parameter and PKCE parameter using Auth
Hi
I am trying to setup STATE and PKCE parameters. I am trying to use the `getState`, `setState`, `getPKCE` and `setPKCE`. I am using cognito hosted UI.

But I am not sure from where I could access those functions. Is there a documentation which explains those things.

Is there a function in Auth which could be used for this purpose.

I tried to access getState and setState but the angular app shows error.
`import { getState, setState } from '@aws-amplify/auth/src/OAuth/oauthStorage';`

**Error:**
`ERROR in ./node_modules/@aws-amplify/auth/src/OAuth/oauthStorage.ts
Module build failed (from ./node_modules/@ngtools/webpack/src/index.js):
Error: /home/ubuntu/Documents/ui-service/node_modules/@aws-amplify/auth/src/OAuth/oauthStorage.ts is missing from the TypeScript compilation. Please make sure it is in your tsconfig via the 'files' or 'include' property.
The missing file seems to be part of a third party library. TS files in published libraries are often a sign of a badly packaged library. Please open an issue in the library repository to alert its author and ask them to package the library using the Angular Package Format (https://goo.gl/jB3GVv).`

### PR DESCRIPTION
**Pull request by sammartinez**
*Issue #, if available:*
Fixes #3188
Fixes #3194
Fixes #3177
Fixes #3125
Fixes #3569

*Description of changes:*
This will allow customers to create a custom state and pass it through the federated Sign In method. Docs PR #[770](https://github.com/aws-amplify/docs/pull/770)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/OAuth/OAuth.ts [hunk 1]
```diff
@@ -66,10 +66,13 @@ export default class OAuth {
     domain: string,
     redirectSignIn: string,
     clientId: string,
-    provider: CognitoHostedUIIdentityProvider | string = CognitoHostedUIIdentityProvider.Cognito) {
+    provider: CognitoHostedUIIdentityProvider | string = CognitoHostedUIIdentityProvider.Cognito,
+    customState?: string) {
 
     const state = this._generateState(32);
-    oAuthStorage.setState(state);
+    const isCustomStatePresent = customState ? `${state}-${customState}` : state;
```
Comment (manueliglesias) [inline packages/auth/src/OAuth/OAuth.ts]: Can you rename this? with its current name it gives the impression of being a boolean while it is not.

Comment (sammartinez) [reply packages/auth/src/OAuth/OAuth.ts]: done

Comment (sammartinez) [reply packages/auth/src/OAuth/OAuth.ts]: Actually I changed my mind, Im gonna rename `state` to generatedState` and then use `state` as the name for the `combinedState` variable so we can use the short hand later on

**Comment by stevearoonie** [comment]
Hi there,

Thanks for doing this work. 

I just happened to be trying out Amplify with SSO today so I thought I'd try this PR out by adding a customState of '/activities' to use as a redirect after login. I ran into one error - validateState ended up comparing the incoming '%2Factivities' from the sign in redirect with '/activities' from oAuthStorage. I tried passing in '%2Factivities' as customState instead but this just ended up getting encoded twice. The fix that worked was to also encode the state when storing it in oAuthStorage in OAuth.oauthSignIn, e.g. `oAuthStorage.setState(encodeURIComponent(state));`

**Pull request merged**